src/dbnode/storage/bootstrap/bootstrapper/fs/source.go

```diff
@@ -821,11 +825,56 @@ func (s *fileSystemSource) read(
 		Span:                      span,
 		NowFn:                     s.nowFn,
 	})
-	bootstrapFromDataReadersResult := s.bootstrapFromReaders(run, md,
-		accumulator, runOpts, readerPool, readersCh, builder)
+
+	bootstrapFromReadersRunResult := newRunResult()
+
+	var buildWg sync.WaitGroup
+	for i := 0; i < indexSegmentConcurrency; i++ {
+		alloc := s.opts.ResultOptions().IndexDocumentsBuilderAllocator()
+		segBuilder, err := alloc()
+		if err != nil {
+			return nil, err
+		}
+
+		builder := result.NewIndexBuilder(segBuilder)
+
+		indexOpts := s.opts.IndexOptions()
+		compactor, err := compaction.NewCompactor(indexOpts.DocumentArrayPool(),
+			index.DocumentArrayPoolCapacity,
+			indexOpts.SegmentBuilderOptions(),
+			indexOpts.FSTSegmentOptions(),
+			compaction.CompactorOptions{
+				FSTWriterOptions: &fst.WriterOptions{
+					// DisableRegistry is set to true to trade a larger FST size
+					// for a faster FST compaction since we want to reduce the end
+					// to end latency for time to first index a metric.
+					DisableRegistry: true,
+				},
+			})
+		if err != nil {
+			return nil, err
+		}
+
+		persistManager, err := fs.NewPersistManager(s.opts.FilesystemOptions())
+		if err != nil {
+			return nil, err
+		}
+
+		buildWg.Add(1)
+		go func() {
+			s.bootstrapFromReaders(run, md,
+				accumulator, runOpts, bootstrapFromReadersRunResult,
+				readerPool, readersCh, builder,
+				&bootstrapper.SharedPersistManager{Mgr: persistManager},
+				&bootstrapper.SharedCompactor{Compactor: compactor})
```

Comment:
So this will use the same reader channel but have multiple goroutines pulling from it, correct? Just trying to get an idea of how this ends up actually splitting by `indexSegmentConcurrency`

Reply:
Correct, all the work happens in bootstrapFromReaders, the open readers channel is just so there's no FS open lag associated with starting the next segment of work.